tap_zendesk_chat/streams.py

```diff
@@ -313,28 +346,74 @@ def _sync(self, ctx, *, full_sync):
-        full_sync = self._should_run_full_sync(ctx)
-        self._sync(ctx, full_sync=full_sync)
+class AgentTimeline(IncrementalStream):
+    LIMIT = 1000
+    def _start_request(self, ctx, start_time_microseconds: int):
+        # https://developer.zendesk.com/api-reference/live-chat/chat-api/incremental_export/#parameters
+        params = {
+            "start_time": start_time_microseconds,
+            "limit": self.LIMIT
+        }
+        return self._request(ctx, params=params)
+
+    def _sync(self, ctx: Context, full_sync: bool=False):
+        pull_id = uuid.uuid4()
+
+        url_offset_key = [self.tap_stream_id, "offset", "agent_timeline-incremental.next_url"]
+        ts_bookmark_key = [self.tap_stream_id, "agent_timeline-incremental.end_time"]
+
         if full_sync:
-            ctx.state["chats_last_full_sync"] = ctx.now.isoformat()
+            sync_days, _ = self._get_sync_days_and_last_sync(ctx)
+            start_dt: 'datetime' = dt_parse(ctx.config['start_date']) - timedelta(days=sync_days)
+            ctx.set_bookmark(ts_bookmark_key, start_dt)
+        else:
+            start_time: str = ctx.update_start_date_bookmark(ts_bookmark_key)
+            start_dt: 'datetime' = dt_parse(start_time)
+
+        start_time_microseconds: int = int(start_dt.timestamp()) * MILLISECOND_TO_MICROSECOND_MULTIPLE 
+
+        max_bookmark: 'datetime' = start_dt
+        end_dt: 'datetime' = ctx.now
+
+        end_time = start_dt # This will be moved forward with each request
+        log_attrs = [f"start_dt={start_dt}", f"end_dt={end_dt}"]
+
+        resp = None
+        next_url = None
+        record_counts = []
+
+        while end_time < end_dt:
+            if next_url:
+                resp = self._request(ctx, url=next_url)
+            else:
+                resp = self._start_request(ctx, start_time_microseconds)
+
+            record_count = resp['count']
+            record_counts.append(record_count)
+
+            next_url = resp["next_page"]
+            ctx.set_bookmark(url_offset_key, next_url)
+
+            LOGGER.info('{} Request completed: {}'.format(pull_id, ",".join(log_attrs + [f'count={record_count}'])))
+
+            end_time_microseconds = int(resp["end_time"])
+            end_time: 'datetime' = datetime.fromtimestamp(end_time_microseconds / MILLISECOND_TO_MICROSECOND_MULTIPLE)
+            max_bookmark: 'datetime' = max(max_bookmark, end_time)
+
+            agent_timeline: List[dict] = resp['agent_timeline']
+
+            self.write_page(ctx, agent_timeline)
+            ctx.set_bookmark(ts_bookmark_key, max_bookmark)
             ctx.write_state()
 
+            if not next_url:
+                LOGGER.info('{} Finished syncing: {}'.format(pull_id, ",".join(log_attrs)))
+                break
+
+            if self.should_give_up(pull_id, ctx, record_counts):
+                LOGGER.info('{} Got too few results; giving up on agent_timeline: {}'.format(pull_id, ",".join(log_attrs)))
```

Comment:
I tried with `is_tapering_off` logic, but I didn't get any success with it since it finishes early when full syncing `full_sync_days` ago.